packages/core-data/src/actions.js

```diff
@@ -543,6 +580,73 @@ export function* saveEntityRecord(
 	}
 }
 
+/**
+ * Runs multiple core-data actions at the same time using one API request.
+ *
+ * Example:
+ *
+ * ```
+ * const [ savedRecord, updatedRecord, deletedRecord ] =
+ *   await dispatch( 'core' ).__experimentalBatch( [
+ *     ( { saveEntityRecord } ) => saveEntityRecord( 'root', 'widget', widget ),
+ *     ( { saveEditedEntityRecord } ) => saveEntityRecord( 'root', 'widget', 123 ),
+ *     ( { deleteEntityRecord } ) => deleteEntityRecord( 'root', 'widget', 123, null ),
+ *   ] );
+ * ```
+ *
+ * @param {Array} requests Array of functions which are invoked simultaneously.
+ *                         Each function is passed an object containing
+ *                         `saveEntityRecord`, `saveEditedEntityRecord`, and
+ *                         `deleteEntityRecord`.
+ *
+ * @return {Promise} A promise that resolves to an array containing the return
+ *                   values of each function given in `requests`.
+ */
+export function* __experimentalBatch( requests ) {
+	const batch = createBatch();
+	const dispatch = yield getDispatch();
+	const api = {
+		saveEntityRecord( kind, name, record, options ) {
+			return batch.add( ( add ) =>
+				dispatch( 'core' ).saveEntityRecord( kind, name, record, {
+					...options,
+					__unstableFetch: add,
+				} )
+			);
+		},
+		saveEditedEntityRecord( kind, name, recordId, options ) {
+			return batch.add( ( add ) =>
+				dispatch( 'core' ).saveEditedEntityRecord(
+					kind,
+					name,
+					recordId,
+					{
+						...options,
+						__unstableFetch: add,
+					}
+				)
+			);
+		},
+		deleteEntityRecord( kind, name, recordId, query, options ) {
+			return batch.add( ( add ) =>
+				dispatch( 'core' ).deleteEntityRecord(
+					kind,
+					name,
+					recordId,
+					query,
+					{
+						...options,
+						__unstableFetch: add,
+					}
+				)
+			);
+		},
+	};
+	const results = requests.map( ( request ) => request( api ) );
+	yield __unstableAwaitPromise( batch.run() );
+	return yield __unstableAwaitPromise( Promise.all( results ) );
```

Comment:
By adding some delay (caused by `await batch.run()`) between creating the `results` promises and awaiting them (in other words, adding `.then` and `.catch` handlers), you are opening your code to the following issue:
```js
const p = Promise.reject( 'myerror' );
setTimeout( () => {
  p.catch( error => {
    console.log( 'p failed with:', error );
  } );
}, 1000 );
```
Even though the `p` error is always caught and reported, you'll get a console warning about uncaught promise rejection:
```
Uncaught (in promise) myerror
```
Because the `catch` handler was attached too late. It wasn't present at the time when the promise rejected.

I don't know how to solve this in this specific case. Just wanted to point out that these warnings will happen.

Reply:
Hmmm. I don't know enough about how `Promise` detects uncaught errors. Would something like this work?

```js
const [ , ...results ] = yield __unstableAwaitPromise(
	Promise.all( [
		batch.run(),
		...requests.map( ( request ) => request( api ) ),
	] )
);
return results;
```

Reply:
@noisysocks the warning will happen when the rejection happen before error handler is bound; your example would probably work

Reply:
👍 23bc3f2d6b